src/dos/dos_keyboard_layout.cpp

```diff
@@ -958,306 +1021,269 @@ KeyboardErrorCode KeyboardLayout::ReadCodePageFile(const char *requested_cp_file
 	}
 
-	return result;
-}
+	auto new_layout = std::make_unique<KeyboardLayout>();
 
-KeyboardErrorCode DOS_SwitchKeyboardLayout(const char* new_layout, int32_t& tried_cp)
-{
-	if (loaded_layout) {
-		KeyboardLayout *changed_layout = nullptr;
-		const auto rcode = loaded_layout->SwitchKeyboardLayout(new_layout, changed_layout, tried_cp);
-		if (changed_layout) {
-			// Remove old layout, activate new layout
-			loaded_layout.reset(changed_layout);
+	std::vector<uint16_t> code_pages = {};
+
+	if (!IS_EGAVGA_ARCH) {
+		// Can't set custom code page on pre-EGA display adapters
+		if (code_page_autodetect) {
+			code_pages.push_back(DefaultCodePage);
+		} else if (code_page != DefaultCodePage || !cpi_file.empty()) {
+			return KeyboardLayoutResult::IncompatibleMachine;
 		}
-		return rcode;
-	} else
-		return KEYB_LAYOUTNOTFOUND;
-}
+	} else if (code_page_autodetect) {
+		// Get the preferred code page
+		const auto bundled = DOS_GetBundledCodePage(keyboard_layout);
+		code_pages = DOS_SuggestBetterCodePages(bundled, keyboard_layout);
+		code_pages.push_back(bundled);
+	} else {
+		code_pages.push_back(code_page);
+	}
 
-// get currently loaded layout name (nullptr if no layout is loaded)
-const char* DOS_GetLoadedLayout()
-{
-	if (loaded_layout) {
-		return loaded_layout->GetLayoutName();
+	// Try to read the keyboard layout
+	KeyboardLayoutResult result = KeyboardLayoutResult::LayoutNotKnown;
+	for (const auto try_code_page : code_pages) {
+		code_page = try_code_page;
+		result = new_layout->ReadKeyboardFile(keyboard_layout, code_page);
+		if (result == KeyboardLayoutResult::OK) {
+			break;
+		}
 	}
-	return nullptr;
-}
 
-// A helper that loads a layout given only a language
-KeyboardErrorCode DOS_LoadKeyboardLayoutFromLanguage(const char * language_pref)
-{
-	assert(language_pref);
+	// Allow using 'us' layout with any code page
+	if ((result == KeyboardLayoutResult::LayoutNotKnown ||
+	     result == KeyboardLayoutResult::NoLayoutForCodePage) &&
+	    iequals(keyboard_layout, "US")) {
+		// Allow using 'us' layout with any code page
+		result = new_layout->ReadKeyboardFile(keyboard_layout,
+		                                      DefaultCodePage);
+	}
 
-	// If a specific language wasn't provided, get it from setup
-	std::string language = language_pref;
-	if (language == "auto") {
-		language = control->GetLanguage();
+	// If no suitable code page found, terminate
+	if (result != KeyboardLayoutResult::OK) {
+		if (result == KeyboardLayoutResult::NoLayoutForCodePage &&
+		    code_page_autodetect) {
+			return KeyboardLayoutResult::IncompatibleMachine;
+		}
+		return result;
 	}
 
-	// TODO: This code mixes language code with keyboard layout; this should
-	//       be cleaned up eventually, possibly we should use
-	//       'use get_language_from_os()' from 'cross.h'
+	// Keyboard layout loaded successfully, change code page accordingly
+	if (!IS_EGAVGA_ARCH || (prefer_rom_font && code_page == DefaultCodePage)) {
```

Comment:
I'd prefer (machine == MCH_EGA) || (machine == MCH_VGA) in new code.

Reply:
Changed in the whole file.